pkg/pipeline/hooks/drivers/bitbucket.go

```diff
@@ -0,0 +1,143 @@
+package drivers
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/rancher/rancher/pkg/pipeline/providers"
+	"github.com/rancher/rancher/pkg/pipeline/remote/bitbucketcloud"
+	"github.com/rancher/rancher/pkg/pipeline/remote/model"
+	"github.com/rancher/rancher/pkg/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	BitbucketCloudWebhookHeader  = "X-Hook-UUID"
+	bitbucketCloudEventHeader    = "X-Event-Key"
+	bitbucketCloudPushEvent      = "repo:push"
+	bitbucketCloudPrCreatedEvent = "pullrequest:created"
+	bitbucketCloudPrUpdatedEvent = "pullrequest:updated"
+	bitbucketCloudStateOpen      = "OPEN"
+)
+
+type BitbucketCloudDriver struct {
+	PipelineLister             v3.PipelineLister
+	PipelineExecutions         v3.PipelineExecutionInterface
+	SourceCodeCredentials      v3.SourceCodeCredentialInterface
+	SourceCodeCredentialLister v3.SourceCodeCredentialLister
+}
+
+func (b BitbucketCloudDriver) Execute(req *http.Request) (int, error) {
+	event := req.Header.Get(bitbucketCloudEventHeader)
+	if event != bitbucketCloudPushEvent && event != bitbucketCloudPrCreatedEvent && event != bitbucketCloudPrUpdatedEvent {
+		return http.StatusUnprocessableEntity, fmt.Errorf("not trigger for event:%s", event)
+	}
+
+	pipelineID := req.URL.Query().Get("pipelineId")
+	ns, name := ref.Parse(pipelineID)
+	pipeline, err := b.PipelineLister.Get(ns, name)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return http.StatusUnprocessableEntity, err
+	}
+	logrus.Infof("get body:%s", string(body))
```

Comment:
don't we have to close the body?

Reply:
As the comments in http.Request suggest,
>// The Server will close the request body. The ServeHTTP
	// Handler does not need to.